src/coreComponents/constitutive/relativePermeability/TableRelativePermeabilityHysteresis.cpp

```diff
@@ -622,6 +626,24 @@ void TableRelativePermeabilityHysteresis::saveConvergedPhaseVolFractionState( ar
 
 }
 
+
+void TableRelativePermeabilityHysteresis::updateTrappedPhaseVolFraction( arrayView2d< real64 const, compflow::USD_PHASE > const & phaseVolFraction ) const
+{
+
+  arrayView2d< real64, compflow::USD_PHASE > phaseTrapped = m_phaseTrapped.toView();
+  arrayView2d< real64, compflow::USD_PHASE > phaseCrit = m_phaseCriticalVolFraction.toView();
+
+  localIndex const numElems = phaseVolFraction.size( 0 );
+  integer const numPhases = numFluidPhases();
+  forAll< parallelDevicePolicy<> >( numElems, [=] GEOSX_HOST_DEVICE ( localIndex const ei ) {
+    for( integer ip = 0; ip < numPhases; ++ip )
+    {
+      phaseTrapped[ei][ip] = LvArray::math::min( phaseVolFraction[ei][ip],
+                                                 phaseCrit[ei][ip] );
+    }
+  } );
+}
```

Comment:
I have a suggestion to save an `array2d` (but I may be wrong). 

Currently:
- You compute `phaseCriticalVolFraction` in the hysteretic kernels
- At the end of the time step, you call `updateTrappedPhaseVolFraction` to take the min of `phaseVolFraction` and `phaseCriticalVolFraction`

Instead, would it be possible to do the following:
- You remove `phaseCriticalVolFraction` from the classes
- You pass `phaseTrappedVolFraction` to the hysteretic kernels
- Inside them, you compute and use `Scrt` as before
- At the end of the hysteretic kernels, you do: `phaseTrappedVolFraction = min( S, Scrt )`
- You remove `updateTrappedPhaseVolFraction`

but I may be missing something.

Reply:
Great idea ! should be done in https://github.com/GEOSX/GEOSX/pull/2052/commits/8052c80a2dc45b9df69c7c4f5b22a0d51aa127ad.

It might not be ideal as it implies quite some code duplication over the ``compute()`` functions of all relperm classes.
I tried though to have ``compute()`` pure virtual in ``relativePermeabilityBase`` instead of ``update`` and use update to trigger ``compute()`` and ``updateTrappedVolFrac()`` but bump into the change of interface of ``compute()`` for ``TableRelativePermeabilityHysteresis``.  That would imply using some adaptation I guess. 

We might save that for future refactoring if the idea was worth pursuing.